src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergColumnTreeFactory.java

```diff
@@ -0,0 +1,169 @@
+package com.snowflake.kafka.connector.internal.streaming.schemaevolution.iceberg;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+
+public class IcebergColumnTreeFactory {
+
+  private final KCLogger LOGGER = new KCLogger(IcebergColumnTreeFactory.class.getName());
+
+  private final IcebergColumnTypeMapper mapper;
+
+  public IcebergColumnTreeFactory() {
+    this.mapper = new IcebergColumnTypeMapper();
+  }
+
+  IcebergColumnTree fromIcebergSchema(IcebergColumnSchema columnSchema) {
+    LOGGER.debug(
+        "Attempting to parse schema from schema stored in a channel for column: "
+            + columnSchema.getColumnName());
+    IcebergFieldNode rootNode =
+        createNode(columnSchema.getColumnName().toUpperCase(), columnSchema.getSchema());
+    return new IcebergColumnTree(rootNode);
+  }
+
+  IcebergColumnTree fromJson(IcebergColumnJsonValuePair pair) {
+    LOGGER.debug(
+        "Attempting to parse schema from records payload for column: " + pair.getColumnName());
+    IcebergFieldNode rootNode = createNode(pair.getColumnName().toUpperCase(), pair.getJsonNode());
+    return new IcebergColumnTree(rootNode);
+  }
+
+  IcebergColumnTree fromConnectSchema(Field kafkaConnectField) {
```

Comment:
We have two separated flows: JSON without schema and AVRO/Protobuf.

I would extract `fromConnectSchema` to a separate class `SchematizedIcebergColumnTreeFactory` and rename this class to `NoSchemaIcebergColumnTreeFactory`. WDYT?

Reply:
Hmm, so what we have:
* from Iceberg (channel) schema - it's used before **both** workflows
* from json payload - workflow without schema
* from record schema - now there is only fromConnectSchema method. Do we have to also write seperate logic to parse avro and protobuf schema? I think not because Converter will parse it into connect schema when we ancounter AVRO or Protobuf. (For sure we ust test it)

When I wrote the factory, I thought it will be a bit over engineering to split it. However if we are going to need more methods then sure. Logically we have 3 parts.